bin/oref0-pump-loop.sh

```diff
@@ -883,16 +846,51 @@ function check_tempbasal() {
   set -o pipefail
   mdt tempbasal 2>&3 | tee monitor/temp_basal.json >&4 && cat monitor/temp_basal.json | jq .temp >&4
 }
+
+# clear and refresh the 24h pumphistory file approximatively every 6 hours.
+# It queries 27h of data, full refresh when oldest data is greater than 33 hours old.
+function pumphistory_daily_refresh() {
+  lastRecordTimestamp=$(jq -r '.[-1].timestamp' monitor/pumphistory-24h-zoned.json 2>&3)
+  dateCutoff=$(date --date="33 hours ago" +%s)
+  echo "Daily refresh if $lastRecordTimestamp < $dateCutoff " >&3
+  if [[ -z "$lastRecordTimestamp" || "$lastRecordTimestamp" == *"null"* || "$(date -d $lastRecordTimestamp +%s)" -le $dateCutoff ]]; then
+    echo -n "daily " && read_full_pumphistory
+  fi
+}
 function read_pumphistory() {
-    read_pumphistory_24h
-  #set -o pipefail
-  #pumphistory -n 1 2>&3 | jq -f openaps.jq | tee monitor/pumphistory-zoned.json 2>&3 >&4 \
-    #&& cat monitor/pumphistory-zoned.json | jq .[0].timestamp
+  set -o pipefail
+  topRecordTimestamp=$(jq -r '.[0].timestamp' monitor/pumphistory-24h-zoned.json 2>&3)
+  echo "Quering pump for history since: $topRecordTimestamp" >&3
+  if [[ -z "$topRecordTimestamp" || "$topRecordTimestamp" == *"null"* ]]; then
+    read_full_pumphistory
+  else
+    # FIXME: the following logic queries the pump for all records since the
+    # timestamp of the top record in the existing history file.
+    # This might miss some records if the pump clock has been moved forward.
+    # A better approach might to get all reconds until that top record
+    # has been found and matched exactly by it's base64 data or some other identifier
+    # other than the timestamp.
+    # The logic could be improved once the pumphistory command support this feature.
+    echo -n "Pump history update"
+    try_fail mv monitor/pumphistory-24h-zoned.json monitor/pumphistory-24h-zoned-old.json
+    if jq -s '.[0] + .[1]' <(pumphistory -s $topRecordTimestamp  2>&3 | jq -f openaps.jq 2>&3 ) monitor/pumphistory-24h-zoned-old.json > monitor/pumphistory-24h-zoned.json; then
+        try_fail rm monitor/pumphistory-24h-zoned-old.json
+        echo -n "d through $(jq -r '.[0].timestamp' monitor/pumphistory-24h-zoned.json); "
+    else
+        try_fail mv monitor/pumphistory-24h-zoned-old.json monitor/pumphistory-24h-zoned.json
+        echo " failed. Last record $(jq -r '.[0].timestamp' monitor/pumphistory-24h-zoned.json)"
+        return 1
+    fi
+  fi
 }
-function read_pumphistory_24h() {
+function read_full_pumphistory() {
   set -o pipefail
-  pumphistory -n 27 2>&3 | jq -f openaps.jq | tee settings/pumphistory-24h-zoned.json 2>&3 >&4 \
-    && cat settings/pumphistory-24h-zoned.json | jq .[0].timestamp
+  rm monitor/pumphistory-24h-zoned.json
+  echo -n "Full history refresh" \
+  && ((( pumphistory -n 27 2>&3 | jq -f openaps.jq 2>&3 | tee monitor/pumphistory-24h-zoned.json 2>&3 >&4 ) \
```

Comment:
Maybe this can be structured the same way as `read_fullpumphistory` using tryfail?

Reply:
Actually there might be a small issue with this code. This just had happen to me luckily while i was semi-monitoring the logs and the pumphistory file size. 

My rig was on another floor with poor reception, and the daily full history refresh failed leaving the pumphistory file empty. It corrected on the next loop, but i'm wondering if we should save the result on a sepearete file and only overwrite on sucess?

Reply:
Yeah, that would probably be slightly better. Should be safe either way: as long as the pumphistory is left either old or empty it won’t loop, but for reporting leaving the old file unchanged is better.